File: Sources/TePlannerKit/Capabilities/Tesla/Security.swift
import Foundation

/// `tesla.security.set_sentry` — toggle sentry mode. Marked
/// `safetyClass = .security` because turning sentry off weakens the
/// parked-car security posture and any user-authored rule invoking it
/// needs an explicit acknowledge.
public struct SetSentryModeCapability: Capability {
    public let id = "tesla.security.set_sentry"
    public let safetyClass: SafetyClass = .security

    public init() {}

    public func invoke(
        ctx: CapabilityContext,
        params: [String: JSONValue],
        api: APIServiceProtocol
    ) async -> CapabilityResult {
        guard let on = params.bool("on") else {
            return CapabilityResult(success: false, error: "on must be boolean")
        }
        guard let vehicleId = ctx.vehicleId else {
            return CapabilityResult(success: false, error: "vehicleId required")
        }
        let result = await api.setSentryMode(vehicleId: vehicleId, on: on)
        switch result {
        case .success:
            return CapabilityResult(success: true)
        case .failure(let err):
            return CapabilityResult(success: false, error: err.localizedDescription)
        }
    }
}

File: backend/app/services/capabilities/tesla/security.py
"""Tesla security-related capabilities (sentry, locks)."""

from __future__ import annotations

from app.services.capabilities import register
from app.services.capabilities.base import (
    Capability,
    CapabilityCallContext,
    CapabilityResult,
    SafetyClass,
)


class SetSentryMode(Capability):
    """Toggle sentry mode. Disabling weakens parked-car security
    posture, so user-authored rules invoking this need explicit
    confirmation."""

    @property
    def id(self) -> str:
        return "tesla.security.set_sentry"

    @property
    def safety_class(self) -> SafetyClass:
        return SafetyClass.SECURITY

    @property
    def params_schema(self) -> dict:
        return {
            "type": "object",
            "properties": {"on": {"type": "boolean"}},
            "required": ["on"],
        }

    async def invoke(
        self, ctx: CapabilityCallContext, params: dict
    ) -> CapabilityResult:
        on = params.get("on")
        if not isinstance(on, bool):
            return CapabilityResult(success=False, error="on must be boolean")
        if not ctx.vin:
            return CapabilityResult(success=False, error="VIN required")
        await ctx.tesla_client.set_sentry_mode(ctx.vin, on)
        return CapabilityResult(success=True, data={"on": on})

    def expected_state(self, params: dict) -> dict:
        on = params.get("on")
        if not isinstance(on, bool):
            return {}
        return {"vehicle.sentry_mode_on": on}


register(SetSentryMode())
